src/common-libs/auth/permission-decorator.ts

```diff
@@ -0,0 +1,47 @@
+import * as auth from '../auth';
+import { createMethodDecorator, ResolverData } from 'type-graphql';
+import Context from '../../domain-services/Context';
+import {
+  RequiredPermissionsCondition,
+  RequiredPermissionsConditionWithId,
+} from './permissions';
+
+// eslint-disable-next-line @typescript-eslint/naming-convention
+export function Permission(
+  requiredPermissions: RequiredPermissionsCondition
+): MethodDecorator {
+  return createMethodDecorator(
+    async ({ args, context }: ResolverData<Context>, next) => {
+      let permissions: RequiredPermissionsConditionWithId;
+
+      if (
+        requiredPermissions !== 'anyone' &&
+        requiredPermissions !== 'noone' &&
+        requiredPermissions.type !== 'global'
+      ) {
+        if (!args.id) {
+          throw new Error('Non global permissions require ID field provided');
+        }
+
+        permissions = {
+          id: args.id,
+          ...requiredPermissions,
+        };
+      } else {
+        permissions = requiredPermissions;
+      }
```

Comment:
I don't think I like this construction of permissions, for a number of reasons:

* It assumes that `args.id` will always be the ID of the target we want to authorize against
* We can only ever have 1 permission, with 1 ID (what if we wanted to authenticate against a governingEntity, and needed to allow either a grant for the governingEntity, or the plan that it belongs to).
* We could have better type-checking guarantees.

As an alternatve, I suggest that you make `requiredPermissions` a function that takes `ResolverData<Context>` as an argument, and returns a `Promise<RequiredPermissionsCondition>`, which you then call inside the body of the decorator. That allows the ID etc... to be precisely determined (in a type-checked manner), and also allows for the re-introduction of more advanced permission conditions,

Reply:
I accepted your suggestion to create a function that takes `ResolverData<Context>` as an argument, and returns a `Promise<RequiredPermissionsCondition>`.